text/trace/0000-error_flagging.md

```diff
@@ -0,0 +1,89 @@
+# Error Flagging with Status Codes
+This proposal adds two status codes explicitly for use as overrides by the end user, and proposes a canonical mapping of semantic conventions to status codes. This clarifies how error reporting should work in OpenTelemetry.
+ 
+## Motivation
+Error reporting is a fundamental use case for distributed tracing. While we prefer that error flagging occurs within analysis tools, and not within instrumentation plugins, a number of currently supported analysis tools and protocols rely on the existence of an explicit error flag reported from instrumentation. In OpenTelmetry, the error flag is called "status codes."
+ 
+However, there is confusion over the mapping of semantic conventions to status codes, and concern over the subjective nature of errors. Which network failures count as an error? Are 404s an error? The answer is dependent on the situation.
+ 
+There is one major exception. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate both when a span should and should not be counted as an error.
+A second exception is supporting analysis tools which require explicit error flagging in the data which they receive. In this case, an operator must be able to apply an error flagging schema at some point during the OTLP data processing pipeline.
+ 
+## Explanation
+The following changes add several missing features required for proper error reporting, and are completely backwards compatible with OpenTelemetry today.
+ 
+### Status Codes
+The following status codes are added to our current  schema.
+ 
+* `DEFAULT` No status has been set. Any errors must be detected by the analysis tool. (This replaces `OK`.)
+* `ERROR` Instrumentation has marked a span as an error. (This replaces `UNKNOWN`.)
+* `OK_OVERRIDE` The user has provided an override. The span should NOT be flagged as an error, regardless of other analysis.
+* `ERROR_OVERRIDE` The user has provided an override. The span SHOULD be flagged as an error, regardless of other analysis.
+ 
+(Note that our  current status codes include a long list of error types. We may choose to keep  them, change them, or drop them in favor of a single `ERROR` code. How many error types we have is not relevant to this proposal .)
+ 
+`OK_OVERRIDE` and `ERROR_OVERRIDE` are special status codes. These are explicit overrides provided by the end user, and should never be set by shared instrumentation. They should only be set by the application developer (via application code), or by the operator (via the collector).
```

Comment:
How would an application developer set the status codes provided by shared instrumentation?

Reply:
Possibly these overrides should be a separate field, so that we retain that data.

Reply:
I initially agreed with the idea of splitting these into two fields. Then it occured to me that we might not have matching mental models.  There are two ways we could split these concepts into two fields:

```
  status: ERROR
  user_override: true
```
or
```
  status: ERROR
  user_status: OK
```

The first replaces the status (which is useful, so there's only one field to check in the back-end). The second introduces a new status field, which supercedes the other status field, but it also keeps both statuses, so a back-end could potentially see what the _instrumentation_ thought the status should be.

If one of our guiding principals is that user-provided status is unilaterally more important than instrumentation-provided status, I think the first option makes the most sense.

Reply:
Yeah, I'm fine with the first one as well. This is a good improvement to this proposal, I'll make the change.